### PR TITLE
add reference operations to backend

### PR DESCRIPTION
**Pull request by cseed**
Builds on: https://github.com/hail-is/hail/pull/5161 (approved, should land shortly)

Add reference operations to Backend.  This is a first cut/stop gap.  The backend needs to be per-user stateless, so we should eventually track the references in Python send references along with IR in execute and resolve references during parsing.

Moved defaultReference to Python.

A little cleanup in TestUtils.

This is the last PR of my big service push.  There some additional things to do, I'll create an issue with bullet points to track them.

### CODE REVIEW

File: hail/python/hail/backend/backend.py [hunk 1]
```diff
@@ -41,7 +41,38 @@ def persist_matrix_table(self, mt, storage_level):
 
     def unpersist_matrix_table(self, mt, storage_level):
         return mt
-    
+
+    @abc.abstractmethod
+    def add_reference(self, config):
+        raise NotImplementedError
```
Comment (tpoterba) [inline hail/python/hail/backend/backend.py]: I think people often write either `pass` or `...` instead of `raise NotImplementedError` in these cases

Comment (tpoterba) [reply hail/python/hail/backend/backend.py]: (you can't instantiate a class without overriding them anyway)

Comment (cseed) [reply hail/python/hail/backend/backend.py]: That's not true in my experience, which is why I added the raise.

```
$ cat test.py
import abc

class A(object):
    @abc.abstractmethod
    def foo(self):
        pass

class B(A):
    def goo(self):
        return 5

b = B()
print('foo', b.foo())
print('goo', b.goo())
$ python test.py
foo None
goo 5
```

Comment (cseed) [reply hail/python/hail/backend/backend.py]: Ah, the abstract base has to have ABCMeta as its metaclass for this to get checked.  Will fix.

**Comment by danking** [comment]
This PR introduced a bug where `hl.get_reference('GRCh37')` only works *after* a call to `hl.init()`. I created an issue https://github.com/hail-is/hail/issues/5252

**Pull request merged**